src/wp-includes/load.php

```diff
@@ -1490,6 +1490,96 @@ function wp_doing_cron() {
 	return apply_filters( 'wp_doing_cron', defined( 'DOING_CRON' ) && DOING_CRON );
 }
 
+/**
+ * Determines whether the current request is a REST API request.
+ *
+ * This function's accuracy cannot be guaranteed until the {@see 'parse_request'} hook is executing.
+ * Earlier usage uses an estimation and may not be accurate on highly customized server setups.
+ *
+ * @since 5.7.0
+ *
+ * @return bool True if it's a REST API request, false otherwise.
+ */
+function wp_doing_rest() {
+	if ( defined( 'REST_REQUEST' ) || did_action( 'parse_request' ) ) {
+		$doing_rest = defined( 'REST_REQUEST' ) && REST_REQUEST;
+
+		/** This filter is documented in wp-includes/load.php */
+		return apply_filters( 'wp_doing_rest', $doing_rest );
+	}
+
+	// This intentionally doesn't use wp_using_themes(). We are trying to check if wp() would be called for this request
+	// in a front-end context which is typically indicated by the WP_USE_THEMES constant.
+	$doing_rest = defined( 'WP_USE_THEMES' ) && WP_USE_THEMES && isset( $_GET['rest_route'] );
+
+	if ( ! get_option( 'permalink_structure' ) || $doing_rest ) {
+		/** This filter is documented in wp-includes/load.php */
+		return apply_filters( 'wp_doing_rest', $doing_rest );
+	}
+
+	if ( isset( $GLOBALS['wp_rewrite'] ) ) {
+		$index = $GLOBALS['wp_rewrite']->index;
+	} else {
+		$index = 'index.php';
+	}
+
+	$pathinfo         = isset( $_SERVER['PATH_INFO'] ) ? $_SERVER['PATH_INFO'] : '';
```

Comment:
I know this is a mostly a copy-paste from `WP` but a couple of tiny things could be tidied up along the way.

```suggestion
	$path_info         = isset( $_SERVER['PATH_INFO'] ) ? $_SERVER['PATH_INFO'] : '';
```
:1234: